File: src/main/java/ru/hollowhorizon/hc/common/objects/items/HollowArmor.java
package ru.hollowhorizon.hc.common.objects.items;

import net.minecraft.data.models.blockstates.PropertyDispatch;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.EquipmentSlot;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.item.ArmorItem;
import net.minecraft.world.item.ArmorMaterial;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.registries.DeferredRegister;
import ru.hollowhorizon.hc.client.utils.HollowPack;

public class HollowArmor<T extends ArmorItem> {
    private final T helm;
    private final T chest;
    private final T legs;
    private final T boots;

    public HollowArmor(PropertyDispatch.TriFunction<ArmorMaterial, EquipmentSlot, Item.Properties, T> armor, ArmorMaterial material, Item.Properties properties) {
        
        this.helm = armor.apply(material, EquipmentSlot.HEAD, properties);
        this.chest = armor.apply(material, EquipmentSlot.CHEST, properties);
        this.legs = armor.apply(material, EquipmentSlot.LEGS, properties);
        this.boots = armor.apply(material, EquipmentSlot.FEET, properties);
    }

    public static <T extends ArmorItem> boolean isFullSet(LivingEntity entity, HollowArmor<T> armor) {
        ItemStack head = entity.getItemBySlot(EquipmentSlot.HEAD);
        ItemStack chest = entity.getItemBySlot(EquipmentSlot.CHEST);
        ItemStack legs = entity.getItemBySlot(EquipmentSlot.LEGS);
        ItemStack feet = entity.getItemBySlot(EquipmentSlot.FEET);

        return head.getItem() == armor.helm && chest.getItem() == armor.chest && legs.getItem() == armor.legs && feet.getItem() == armor.boots;
    }

    public static <T extends ArmorItem> void damagePart(LivingEntity entity, EquipmentSlot target, int damage) {
        ItemStack armorItem = entity.getItemBySlot(target);

        armorItem.hurtAndBreak(damage, entity, (entity1 -> {}));
    }

    public static <T extends ArmorItem> boolean hasPart(LivingEntity entity, HollowArmor<T> armor, EquipmentSlot target) {
        ItemStack arm = entity.getItemBySlot(target);

        return arm.getItem() == armor.helm || arm.getItem() == armor.chest || arm.getItem() == armor.legs || arm.getItem() == armor.boots;
    }

    public static <T extends ArmorItem> boolean isContainsAnyPart(LivingEntity entity, HollowArmor<T> armor) {
        ItemStack head = entity.getItemBySlot(EquipmentSlot.HEAD);
        ItemStack chest = entity.getItemBySlot(EquipmentSlot.CHEST);
        ItemStack legs = entity.getItemBySlot(EquipmentSlot.LEGS);
        ItemStack feet = entity.getItemBySlot(EquipmentSlot.FEET);

        return head.getItem() == armor.helm || chest.getItem() == armor.chest || legs.getItem() == armor.legs || feet.getItem() == armor.boots;
    }

    public void registerItems(DeferredRegister<Item> register, String name) {
        register.register(name + "_helmet", () -> helm);
        register.register(name + "_chest", () -> chest);
        register.register(name + "_legs", () -> legs);
        register.register(name + "_boots", () -> boots);
    }

    public void registerModels(String modid, String name) {
        HollowPack.genItemModels.add(new ResourceLocation(modid, name+"_helmet"));
        HollowPack.genItemModels.add(new ResourceLocation(modid, name+"_chest"));
        HollowPack.genItemModels.add(new ResourceLocation(modid, name+"_legs"));
        HollowPack.genItemModels.add(new ResourceLocation(modid, name+"_boots"));
    }

}

File: src/main/java/ru/hollowhorizon/hc/common/objects/items/HollowArmor.kt
package ru.hollowhorizon.hc.common.objects.items

import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.EquipmentSlot
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.item.ArmorItem
import net.minecraft.world.item.ArmorMaterial
import net.minecraft.world.item.Item
import net.minecraftforge.registries.DeferredRegister
import net.minecraftforge.registries.RegistryObject
import ru.hollowhorizon.hc.client.utils.HollowPack

class HollowArmor<T : ArmorItem>(
    armor: (ArmorMaterial, EquipmentSlot, Item.Properties) -> T,
    material: ArmorMaterial,
    properties: Item.Properties,
) {
    val helmet: () -> T = { armor(material, EquipmentSlot.HEAD, properties) }
    val chest: () -> T = { armor(material, EquipmentSlot.CHEST, properties) }
    val legs: () -> T = { armor(material, EquipmentSlot.LEGS, properties) }
    val boots: () -> T = { armor(material, EquipmentSlot.FEET, properties) }

    fun registerItems(register: DeferredRegister<Item>, name: String): RegistryObject<T> {
        register.register(name + "_boots", boots)
        register.register(name + "_legs", legs)
        register.register(name + "_chest", chest)
        return register.register(name + "_helmet", helmet)
    }

    fun registerModels(modid: String, name: String) {
        HollowPack.genItemModels.add(ResourceLocation(modid, name + "_helmet"))
        HollowPack.genItemModels.add(ResourceLocation(modid, name + "_chest"))
        HollowPack.genItemModels.add(ResourceLocation(modid, name + "_legs"))
        HollowPack.genItemModels.add(ResourceLocation(modid, name + "_boots"))
    }

    companion object {
        fun isFullSet(entity: LivingEntity, armor: HollowArmor<*>): Boolean {
            val head = entity.getItemBySlot(EquipmentSlot.HEAD)
            val chest = entity.getItemBySlot(EquipmentSlot.CHEST)
            val legs = entity.getItemBySlot(EquipmentSlot.LEGS)
            val feet = entity.getItemBySlot(EquipmentSlot.FEET)
            return head.item === armor.helmet && chest.item === armor.chest && legs.item === armor.legs && feet.item === armor.boots
        }

        fun damagePart(entity: LivingEntity, target: EquipmentSlot, damage: Int) {
            val armorItem = entity.getItemBySlot(target)
            armorItem.hurtAndBreak(damage, entity) { _ -> }
        }

        fun hasPart(entity: LivingEntity, armor: HollowArmor<*>, target: EquipmentSlot?): Boolean {
            val arm = entity.getItemBySlot(target!!)
            return arm.item === armor.helmet || arm.item === armor.chest || arm.item === armor.legs || arm.item === armor.boots
        }

        fun isContainsAnyPart(entity: LivingEntity, armor: HollowArmor<*>): Boolean {
            val head = entity.getItemBySlot(EquipmentSlot.HEAD)
            val chest = entity.getItemBySlot(EquipmentSlot.CHEST)
            val legs = entity.getItemBySlot(EquipmentSlot.LEGS)
            val feet = entity.getItemBySlot(EquipmentSlot.FEET)
            return head.item === armor.helmet || chest.item === armor.chest || legs.item === armor.legs || feet.item === armor.boots
        }
    }
}
